### PR TITLE
CPE matching 

### PR DESCRIPTION
**Pull request by adamjanovsky**
Implements CPE matching code. This involves:
- Class for `CPEDataset`
- Functions for fuzzy CPE matching
- Functions for manual labeling of CPE matches based on the automatically found candidates
- Class for `CVEDataset`
- Functions for deterministic matching of CVEs to CPE-rich certificates
- Some tests of CPE and CVE classes
- Slight improvement of pandas serialization
- Fixes of encoding errors that resulted in escaping of non-ascii chars in jsons (rendering some links unusable)

## What remains to be done

- [x] The links in serialized json don't work -- encoding issue
- [x] Fix encoding error of csv/html resources in tests
- [x] Add `rapidfuzz` as a dependency into requirements
- [x] Prepare a function to manually validate the findings of the CPE matching
- [x] Add class for handeling CVE dataset
- [x] Add functions to correlate CC certificates with CVEs
- [x] Document the exact matching procedure somewhere
- [x] Add relaxed version of matching when no suitable CPE version is found
- [x] Delete old notebook once done
- [x] Add simple tests for `CPE`
- [x] Add simple tests for `CPEDataset`
- [x] Add simple tests for `CVE`
- [x] Add simple tests for `CVEDataset`

### Requires #52 to be implemented:

- [x] Fine-tune the match threshold
- [x] Compare whether to parse with CPE title or cpe-item field

### Requires #53 to be implemented:

- [x] Store `CVEDataset` along with the `CCDataset`
- [x] Store `CPEDataset` along with the `CPEDataset`

### CODE REVIEW

**Comment by adamjanovsky** [comment]
The CPE matching procedure documented at https://github.com/crocs-muni/sec-certs/wiki/CommonCriteria-certificate-to-CPE-matching-algorithm

**Comment by adamjanovsky** [comment]
@petrs this might be of an interest to you. There's a function that we will use to manually label candidate CPE matches. It's designed so that it handles most of the stuff automatically to keep the labeling as painless as possible. For each of the certificates you get prompted like this:

```
[0/1] Vendor: NetIQ, Incorporated, Name: NetIQ Access Manager 4.5
	- [0]: CPE(uri='cpe:2.3:a:netiq:access_manager:4.5:sp1:*:*:*:*:*:*', title='NetIQ Access Manager 4.5 Service Pack 1', version='4.5', vendor='netiq', item_name='access manager')
	- [1]: CPE(uri='cpe:2.3:a:netiq:access_manager:4.5:hotfix1:*:*:*:*:*:*', title='NetIQ Access Manager 4.5 Hotfix 1', version='4.5', vendor='netiq', item_name='access manager')
	- [2]: CPE(uri='cpe:2.3:a:netiq:access_manager:4.5:-:*:*:*:*:*:*', title='NetIQ Access Manager 4.5', version='4.5', vendor='netiq', item_name='access manager')
	- [X]: No fitting match
Select fitting CPE matches (split with comma if choosing more):
```

and you are expected to provide input, anything of the following will do:
- `1`
- `1,2,3`
- `1, 2, 3`
- 'X'
- 'asdasdsad' -> Will repeat until you answer :)

The progress is periodically save so that you don't have to label all certificates at once. Only the so-far-unlabeled certificates are of course shown.

**Comment by adamjanovsky** [comment]
@J08nY not sure how much you feel for the review. The code probably is not perfect, but if you'd manage to point to major problems, that would be nice. 

I think that you at least want to take a look at [examples](https://github.com/crocs-muni/sec-certs/tree/cpe_matching_notebook/examples) (there's also a readme file) and [cpe matching procedure description](https://github.com/crocs-muni/sec-certs/wiki/CommonCriteria-certificate-to-CPE-matching-algorithm) at wiki.

File: sec_certs/dataset.py [hunk 1]
```diff
@@ -650,34 +681,99 @@ def extract_data(self, fresh: bool = True, update_json: bool = False):
...
+
+                if 'X' not in inpts and 'x' not in inpts:
+                    if 'A' in inpts or 'a' in inpts:
+                        inpts = [x for x in range(0, len(x.heuristics.cpe_matches))]
+                    try:
+                        inpts = [int(x) for x in inpts]
+                        if min(inpts) < 0 or max(inpts) > len(x.heuristics.cpe_matches) - 1:
+                            raise ValueError(f'Incorrect number chosen, choose in range 0-{len(x.heuristics.cpe_matches) - 1}')
+                    except ValueError as e:
+                        logger.error(f'Bad input from user, repeating instance: {e}')
```
Comment (J08nY) [inline sec_certs/dataset.py:747]: Is using `logger` to communicate with the user here a good idea? What if `logger` is redirected somewhere? Logs of the library and user printed output do not mix well in this case. I think just prints or perhaps the use of click here is ok.

File: sec_certs/dataset.py [hunk 2]
```diff
@@ -650,34 +681,99 @@ def extract_data(self, fresh: bool = True, update_json: bool = False):
...
+        certs_to_verify: List[CommonCriteriaCert] = [x for x in self if (x.heuristics.cpe_matches and not x.heuristics.verified_cpe_matches)]
+        logger.info('Manually verifying CPE matches')
+        time.sleep(0.05)  # easier than flushing the logger
+        verify_certs(certs_to_verify)
+
+        if update_json is True:
+            self.to_json()
+
+    def compute_related_cves(self, download_fresh_cves: bool = False):
+        logger.info('Retreiving related CVEs to verified CPE matches')
```
Comment (J08nY) [inline sec_certs/dataset.py:767]: Retr**ei**ving.

**Review by J08nY — CHANGES_REQUESTED** [review]
Overall just requested minor changes w.r.t. some configuration handling, naming of functions and the like.

What I would like to see (but is not in scope of this PR, but this PR touched some of the related thing so I bring it up) is that there are not just "examples" using the new codebase but a full replication of the old command-line scripts, such that someone can use the tool without writing any Python code. AFAIK that is a blocker before removing the old codebase (to reach feature parity and have a simple-to-use CLI tool that produces the JSONs).

File: sec_certs/constants.py [hunk 3]
```diff
@@ -6,6 +6,9 @@
 RETURNCODE_NOK = 'nok'
 REQUEST_TIMEOUT = 10
 
+CPE_MATCHING_THRESHOLD = 70
```
[the remark this answers is not in the document]
Comment (adamjanovsky) [reply sec_certs/constants.py:9]: I agree. Though I'm quite confused with how the config is loaded at the moment. Have to discuss with @mmstanone and do proper config for all parameters of the tool. Will create separate branch/PR for this one.

File: sec_certs/cpe.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            set_match_item = fuzz.token_set_ratio(sanitized_cert_name, sanitized_item_name)
+            partial_match_item = fuzz.partial_ratio(sanitized_cert_name, sanitized_item_name)
+
+            potential = max([set_match_title, partial_match_title, set_match_item, partial_match_item])
+
+            if potential > threshold:
+                reasonable_matches.append((potential, c))
+
+        if reasonable_matches:
+            reasonable_matches = sorted(reasonable_matches, key=lambda x: x[0], reverse=True)
```
[the remark this answers is not in the document]
Comment (adamjanovsky) [reply sec_certs/cpe.py:238]: Nice, did not know this one.

File: sec_certs/dataset.py [hunk 5]
```diff
@@ -144,20 +154,33 @@ def to_dict(self):
         def from_dict(cls, dct: Dict[str, bool]):
             return cls(*tuple(dct.values()))
 
-    def __init__(self, certs: Dict[str, 'Certificate'], root_dir: Path, name: str = 'dataset name',
+    def __init__(self, certs: Dict[str, 'CommonCriteriaCert'], root_dir: Path, name: str = 'dataset name',
                  description: str = 'dataset_description', state: Optional[DatasetInternalState] = None):
         super().__init__(certs, root_dir, name, description)
+        self.certs: Dict[str, 'CommonCriteriaCert'] = certs  # semantically redundant, but the type hint is needed
```
[the remark this answers is not in the document]
Comment (adamjanovsky) [reply sec_certs/dataset.py:160]: [Yep, good tip](https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations)

File: sec_certs/cpe.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        self.vendors = set(self.vendor_to_versions.keys())
+        for cpe in self:
+            self.vendor_to_versions[cpe.vendor].append(cpe.version)
+            if (cpe.vendor, cpe.version) not in self.vendor_version_to_cpe:
+                self.vendor_version_to_cpe[(cpe.vendor, cpe.version)] = [cpe]
+            else:
+                self.vendor_version_to_cpe[(cpe.vendor, cpe.version)].append(cpe)
+
+        # remove duplicates
+        self.vendor_to_versions = {x: list(set(y)) for x, y in self.vendor_to_versions.items()}
```
[the remark this answers is not in the document]
Comment (adamjanovsky) [reply sec_certs/cpe.py:100]: Well, I have no idea why didn't I figure that out myself :).

File: sec_certs/cpe.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        Given parameters, will return Pairs (cpe_vendor, cpe_version) that should are relevant to a given certificate
+        Parameters
+        ----------
+        cert_candidate_cpe_vendors ... list of CPE vendors relevant to a certificate
+        cert_candidate_versions ... List of versions heuristically extracted from the certificate name
+
+        Returns
+        -------
+        List of tuples (cpe_vendor, cpe_version) that can be used in the lookup table to search the CPE dataset.
+        """
```
[the remark this answers is not in the document]
Comment (adamjanovsky) [reply sec_certs/cpe.py:188]: I have the same opinion. Upon new PyCharm install maybe the default docstring style changed and I didn't noticed. Changed it to reStructuredText.

**Comment by adamjanovsky** [comment]
> Overall just requested minor changes w.r.t. some configuration handling, naming of functions and the like.
> 
> What I would like to see (but is not in scope of this PR, but this PR touched some of the related thing so I bring it up) is that there are not just "examples" using the new codebase but a full replication of the old command-line scripts, such that someone can use the tool without writing any Python code. AFAIK that is a blocker before removing the old codebase (to reach feature parity and have a simple-to-use CLI tool that produces the JSONs).

Apart from minor missing stuff (that you pointed out to me), the `cc_oop_demo.py` is actually the full replication of the code. Takes several hours to run and does all of the following:

1. Download CC data from csv and html
2. Download pdf data related to the parsed certificates
3. Convert all pdfs to txt
4. Analyze keywords in all txts
5. Download CPE dataset and compute candidate CPE matches
6. (expert knowledge) manually label CPE-rich certificates
7. Download CVEDataset
8. Correlate CVEs with CPEs

Once I resolve the missing stuff, we're happy to move the old API to some stale branch.

**Comment by adamjanovsky** [comment]
@J08nY thanks for pointing to obvious errors. Fixed them all and merging. 🎉

**Pull request merged**